### PR TITLE
Implement `plot_contour` and `_get_contour_plot` with Matplotlib backend.

### PR DESCRIPTION
**Pull request by ytknzw**
This PR proposes to implement a ``matplotlib`` backend for plotting parameter value contour based on PR #1756 .

## Motivation
To provide another visualization option for users, who have only ``plotly`` now.
See issue #1539 and PR #1707 for the original motivation.

## Description of the changes
Create  ``optuna.visualization.matplotlib._contour.plot_contour``, ``optuna.visualization.matplotlib._contour._get_contour_plot``, ``optuna.visualization.matplotlib._contour._set_cmap``, ``optuna.visualization.matplotlib._contour._calculate_griddata``, ``optuna.visualization.matplotlib._contour._generate_contour_subplot``.

### CODE REVIEW

File: optuna/visualization/matplotlib/_contour.py [hunk 1]
```diff
@@ -1,11 +1,235 @@
...
+        _, ax = plt.subplots()
+        return ax
+
+    all_params = {p_name for t in trials for p_name in t.params.keys()}
+
+    if params is None:
+        sorted_params = sorted(list(all_params))
+    elif len(params) <= 1:
+        _logger.warning("The length of params must be greater than 1.")
+        fig, ax = plt.subplots()
```
Comment (crcrpar) [inline optuna/visualization/matplotlib/_contour.py:67]: [nit-picking]
```suggestion
        _, ax = plt.subplots()
```

**Review by crcrpar — COMMENTED** [review]
Great work and sorry for soooo late response 🙇 

The figure itself looks good.

image
In my environment, the figure of the example code of `optuna.visulaization.plot_contour` is the above which looks a bit different from the one of https://optuna.readthedocs.io/en/latest/reference/visualization/generated/optuna.visualization.plot_contour.html#optuna.visualization.plot_contour.
Is this due to my environment or do you have any idea to improve?

**Comment by ytknzw** [comment]
Thank you for your review!
I should have put a note in PR but nothing's wrong with your environment. It's because of the difference of interpolation algorithms between the Plotly backend and my Matplotlib implementation, and this is what I have achieved with Matplotlib and scipy so far...
I tried to use or recreate the same algorithm Plotly used, but haven't been able to identify it yet. This is why I left a `TODO` in line 180.
https://github.com/ytknzw/optuna/blob/42e679dc6ef1556f8cedb42e70b836a691347c7d/optuna/visualization/matplotlib/_contour.py#L180

**Comment by crcrpar** [comment]
> It's because of the difference of interpolation algorithms between the Plotly backend and my Matplotlib implementation, and this is what I have achieved with Matplotlib and scipy so far...

So my last comment might have sound a bit aggressive to you, but I NEVER meant it. I misunderstood the TODO comment, sorry. 🙇 

How about adding a note and/or warning to the `plot_contour` function since the output figure of this function may differ from that of the plotly implementation?

**Comment by ytknzw** [comment]
> So my last comment might have sound a bit aggressive to you, but I NEVER meant it. I misunderstood the TODO comment, sorry. 🙇

Oh, no, not at all!

> How about adding a note and/or warning to the `plot_contour` function since the output figure of this function may differ from that of the plotly implementation?

Thank you for your suggestion. Added a warning.

**Review by crcrpar — COMMENTED** [review]
Thank you for your adding the warning message. I confirmed it in the below capture. 
image

I have two things to say.

1. Does this support categorical variables? (plotly seems not to emit error.)
2. Better colormap? To get the below contour plot, I created a study whose direction is "maximize" and I could check that the region of better objective value is denser while color is bright, which is not intuitive to me. Is it possible to change the colormap depending on the given study's direction? (I think this can/should be the follow-up).

image

**Comment by ytknzw** [comment]
Thank you for your review.

>  1. Does this support categorical variables? (plotly seems not to emit error.)

Unfortunately, no. I'll add that in this PR.

> 2. Better colormap? To get the below contour plot, I created a study whose direction is "maximize" and I could check that the region of better objective value is denser while color is bright, which is not intuitive to me. Is it possible to change the colormap depending on the given study's direction? (I think this can/should be the follow-up).

Sorry, I think it's my mistake. The direction of Matplotlib's colourmap is opposite to Plotly's. I'll fix it in this PR.

**Review by crcrpar — APPROVED** [review]
Thank you for your reminder. As this's been already so long mainly due to my laziness, I'd love to merge this first and then improve by adding test cases.

**Review by himkt — COMMENTED** [review]
I dropped some minor comments. 😄 

And I'm curious about figures generated by `plotly` and `matplotlib` seems to be different.
I put some examples:

Screen Shot 2020-11-11 at 22 03 23
Screen Shot 2020-11-11 at 22 01 33

It seems for me that surfaces of contour plots are not the same. 🤔 
(code is here: https://colab.research.google.com/drive/1BS3bFWIjCgWSfzmVSELVPswJcvFseCpi#scrollTo=RRfy8tW4Zv4m)

Is this expected behavior?

File: optuna/visualization/matplotlib/_contour.py [hunk 2]
```diff
@@ -1,11 +1,360 @@
...
+_logger = get_logger(__name__)
 
 
 @experimental("2.2.0")
-def plot_contour() -> Axes:
-    raise NotImplementedError("To be implemented soon.")
+def plot_contour(study: Study, params: Optional[List[str]] = None) -> "Axes":
+    """Plot the parameter relationship as contour plot in a study with Matplotlib.
+
+    Note that, If a parameter contains missing values, a trial with missing values is not plotted.
```
[the remark this answers is not in the document]
Comment (ytknzw) [reply optuna/visualization/matplotlib/_contour.py]: Thanks. I lowercased the `If`.

**Comment by ytknzw** [comment]
Thank you for your review.
>It looks for me that surfaces of contour plots are not the same. 🤔
Is this expected behavior?

Yes, it is. It's because of different interpolation algorithms. I haven't been able to identify Plotly's algorithm yet... Please refer to this prior discussion: https://github.com/optuna/optuna/pull/1782#pullrequestreview-507964460
I also corrected Plotly's colour map direction.

**Review by himkt — APPROVED** [review]
> It's because of different interpolation algorithms.

Thank you so much for clarifying! And I'm sorry that I missed the discussion.
So, it looks good to me. Thank you for you continuous contribution!

**Pull request merged**